sdk/python/v1beta1/kubeflow/katib/utils/utils.py

```diff
@@ -34,3 +44,164 @@ def set_katib_namespace(katib):
     katib_namespace = katib.metadata.namespace
     namespace = katib_namespace or get_default_target_namespace()
     return namespace
+
+
+class Deserializer:
```

Comment:
Since after this change: https://github.com/kubeflow/katib/pull/1630#discussion_r697873613, your `__deserialize` is the same as [`__deserialize` from api_client](https://github.com/kubeflow/katib/blob/4a453d8e1e2a4ff390b4691811f0a71c6c2d7366/sdk/python/v1beta1/kubeflow/katib/api_client.py#L282), maybe we could try not to duplicate all of the functions ?
I understand that we can't execute `__deserialize` from the utils, but maybe we can try to do some hack, as here: https://github.com/kubernetes-client/python/issues/977#issuecomment-592030030.

For example, in utils.py follow this:

```
class FakeResponse:
    def __init__(self, obj):
        self.data = obj         ​

```

And in Katib client just do this:

```
katibexp = thread.get(constants.APISERVER_TIMEOUT)
            result = [
                api_client.deserialize(FakeResponse(item), V1beta1Experiment)
                for item in katibexp.get("items")
            ]
```
WDYT @anencore94 ?

Reply:
Wow :open_mouth: , That looks much better!
I feel not clear for copying the whole private deserialize methods :joy: 

I think using some another name rather than `FakeResponse` will be more readable.
Is there any candidates? @andreyvelich

Reply:
We can name it as `RESTResponse`, as an official name of this [response object](https://github.com/kubeflow/katib/blob/4a453d8e1e2a4ff390b4691811f0a71c6c2d7366/sdk/python/v1beta1/kubeflow/katib/api_client.py#L559).
What are your thoughts ?

Reply:
I felt a bit confused we have to specify `kubernetes` or not on naming.
Maybe one of FakeResponse, FakeRESTResponse, FakeKubernetesRESTResponse, ...

Reply:
By the way, we need to use json.dumps : `self.data = json.dumps(obj)` as https://github.com/kubernetes-client/python/issues/977#issuecomment-592030030 did.
Since our response type is already dict, not str or bytes.

Reply:
I think it should work without `dumps`.
Check here: https://github.com/kubeflow/katib/blob/4a453d8e1e2a4ff390b4691811f0a71c6c2d7366/sdk/python/v1beta1/kubeflow/katib/api_client.py#L275-L278.
In case of error: `data = response.data`

Reply:
Let's keep it `FakeResponse`, we can rename it later if it requres.

Reply:
When such case occurs, it raise TypeError, not ValueError

```
Traceback (most recent call last):
  File "/home/kjy/git/katib/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py", line 274, in list_experiments
    for item in katibexp.get("items")
  File "/home/kjy/git/katib/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py", line 274, in <listcomp>
    for item in katibexp.get("items")
  File "/home/kjy/git/katib/sdk/python/v1beta1/kubeflow/katib/api_client.py", line 276, in deserialize
    data = json.loads(response.data)
  File "/home/kjy/.pyenv/versions/3.6.9/lib/python3.6/json/__init__.py", line 348, in loads
    'not {!r}'.format(s.__class__.__name__))
TypeError: the JSON object must be str, bytes or bytearray, not 'dict'

During handling of the above exception, another exception occurred:
```

Reply:
Could we change the **auth-generated** `api_client.py` to catch ValueError and TypeError too ? Or just add skipping parts when `data`'s type is already `dict`

Reply:
Hmm, I am not sure that we can modify this auto-generated file.
Let's keep it `self.data = json.dumps(obj)` as you said before.
Does it deserialize properly in that case ?

Reply:
Yes, with your FakeResponse class and `json.dumps` as in the last commit

```python
    resp = client.list_experiments(namespace='jaeyeon-kim')
    # print(resp)
    print(type(resp[0]))
    print(type(resp[0].metadata))
    print(type(resp[0].spec))
    print(type(resp[0].spec.algorithm))

    resp2 = client.list_trials(name='katib-hpo-test', namespace='jaeyeon-kim')
    # print(resp2[0])
    print(type(resp2[0]))
    print(type(resp2[0].spec.objective))
```
- output : 
```
<class 'kubeflow.katib.models.v1beta1_experiment.V1beta1Experiment'>
<class 'kubernetes.client.models.v1_object_meta.V1ObjectMeta'>
<class 'kubeflow.katib.models.v1beta1_experiment_spec.V1beta1ExperimentSpec'>
<class 'kubeflow.katib.models.v1beta1_algorithm_spec.V1beta1AlgorithmSpec'>
<class 'kubeflow.katib.models.v1beta1_trial.V1beta1Trial'>
<class 'kubeflow.katib.models.v1beta1_objective_spec.V1beta1ObjectiveSpec'>
```

I'm afraid how to add unit-tests for these. I see there isn't any unit-test codes yet for [sdk](https://github.com/kubeflow/katib/tree/master/sdk/python/v1beta1/kubeflow/katib) right ?

Reply:
That's great, thank you!
Please can you also check that `status.startTime` was properly created ? I am worried about this comment: https://github.com/kubernetes-client/python/issues/977#issuecomment-594045477.

Yeah, we don't have tests for the SDK currently: https://github.com/kubeflow/katib/blob/master/hack/gen-python-sdk/post_gen.py#L87.
We can think about it in the following PRs.

> Yes, with your FakeResponse class and json.dumps as in the last commit

Reply:
> Yeah, we don't have tests for the SDK currently: https://github.com/kubeflow/katib/blob/master/hack/gen-python-sdk/post_gen.py#L87.
We can think about it in the following PRs.

Thanks for your support !

Reply:
> Please can you also check that status.startTime was properly created ? I am worried about this comment: kubernetes-client/python#977 (comment).

The output looks good :)
```python
    print(type(resp[0].status))
    print(type(resp[0].status.start_time))
    print(resp[0].status.start_time)
```

- output
```
<class 'kubeflow.katib.models.v1beta1_experiment_status.V1beta1ExperimentStatus'>
<class 'datetime.datetime'>
2021-08-28 12:00:30+00:00
```

Reply:
For make sure, here's the test for `list_trials`
```python
    resp2 = client.list_trials(name='katib-hpo-test', namespace='jaeyeon-kim')
    print(type(resp2[0]))
    print(type(resp2[0].status))
    print(type(resp2[0].status.start_time))
    print(resp2[0].status.start_time)

```

```
<class 'kubeflow.katib.models.v1beta1_trial.V1beta1Trial'>
<class 'kubeflow.katib.models.v1beta1_trial_status.V1beta1TrialStatus'>
<class 'datetime.datetime'>
2021-08-28 12:01:56+00:00
```

Reply:
Great, thank you so much for taking this!

Reply:
It was a long road! Thank you so much for bring me here :+1: